File: cortex-py/cortex/diff.py
"""Diff between two timepoints of agent reasoning."""

from dataclasses import dataclass


@dataclass
class Diff:
    """Diff result between two timepoints.

    Attributes:
        agent_id: Agent identifier
        from_time: Start timestamp
        to_time: End timestamp
        from_count: Number of decisions in first period
        to_count: Number of decisions in second period
        from_avg_confidence: Average confidence in first period
        to_avg_confidence: Average confidence in second period
        confidence_delta: Confidence change (to - from)
        from_tools: Tools used in first period
        to_tools: Tools used in second period
        tools_added: Tools added in second period
        tools_removed: Tools removed in second period
    """

    agent_id: str
    from_time: str
    to_time: str
    from_count: int
    to_count: int
    from_avg_confidence: float
    to_avg_confidence: float
    confidence_delta: float
    from_tools: list[str]
    to_tools: list[str]
    tools_added: list[str]
    tools_removed: list[str]
    from_decision_types: list[str] = None
    to_decision_types: list[str] = None
    from_reasoning_samples: list[str] = None
    to_reasoning_samples: list[str] = None

    def __post_init__(self):
        if self.from_decision_types is None:
            self.from_decision_types = []
        if self.to_decision_types is None:
            self.to_decision_types = []
        if self.from_reasoning_samples is None:
            self.from_reasoning_samples = []
        if self.to_reasoning_samples is None:
            self.to_reasoning_samples = []

    @property
    def summary(self) -> str:
        """Human-readable summary."""
        conf_dir = "+" if self.confidence_delta > 0 else ""
        tool_change = ""
        if self.tools_added or self.tools_removed:
            tool_change = f" Tools: +{len(self.tools_added)} -{len(self.tools_removed)}"

        return (
            f"Decisions: {self.from_count} → {self.to_count}. "
            f"Confidence: {conf_dir}{self.confidence_delta * 100:.1f}%.{tool_change}"
        )

File: cortex-core/src/diff.rs
//! Diff between two timepoints of agent reasoning

use serde::{Deserialize, Serialize};

/// Diff result between two timepoints
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Diff {
    /// Agent ID
    pub agent_id: String,

    /// Start timestamp
    pub from_time: String,

    /// End timestamp
    pub to_time: String,

    /// Number of decisions in first period
    pub from_count: usize,

    /// Number of decisions in second period
    pub to_count: usize,

    /// Average confidence in first period
    pub from_avg_confidence: f32,

    /// Average confidence in second period
    pub to_avg_confidence: f32,

    /// Confidence delta (to - from)
    pub confidence_delta: f32,

    /// Tools used in first period (unique)
    pub from_tools: Vec<String>,

    /// Tools used in second period (unique)
    pub to_tools: Vec<String>,

    /// Tools added (in to, not in from)
    pub tools_added: Vec<String>,

    /// Tools removed (in from, not in to)
    pub tools_removed: Vec<String>,

    /// Decision types in first period
    pub from_decision_types: Vec<String>,

    /// Decision types in second period
    pub to_decision_types: Vec<String>,

    /// Sample reasoning snippets from first period
    pub from_reasoning_samples: Vec<String>,

    /// Sample reasoning snippets from second period
    pub to_reasoning_samples: Vec<String>,
}

impl Diff {
    /// Human-readable summary
    pub fn summary(&self) -> String {
        let conf_dir = if self.confidence_delta > 0.0 { "+" } else { "" };
        let tool_change = if !self.tools_added.is_empty() || !self.tools_removed.is_empty() {
            format!(
                " Tools: +{} -{}",
                self.tools_added.len(),
                self.tools_removed.len()
            )
        } else {
            String::new()
        };

        format!(
            "Decisions: {} → {}. Confidence: {}{:.1}%.{}",
            self.from_count,
            self.to_count,
            conf_dir,
            self.confidence_delta * 100.0,
            tool_change
        )
    }
}
